pkg/volume/csi/shared_informer_hook.go

```diff
@@ -0,0 +1,108 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package csi
+
+import (
+	"sync"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/tools/cache"
+)
+
+// sharedInformerHook provides a way to dynamically watch objects via a SharedInformer.
+type sharedInformerHook struct {
+	informer           cache.SharedIndexInformer
+	objectNamesToWatch map[string]chan watch.Event
+	mu                 sync.RWMutex
+}
+
+func SharedInformerHook(informer cache.SharedIndexInformer) *sharedInformerHook {
+	return &sharedInformerHook{
+		informer:           informer,
+		objectNamesToWatch: make(map[string]chan watch.Event),
+	}
+}
+
+// RegisterHandlerFuncs adds event handlers to the VolumeAttachmentInformer field.
+// Keep in mind this operation is not idempotent and should be called once.
+func (s *sharedInformerHook) RegisterHandlerFuncs() {
+	s.informer.AddEventHandler(
+		&cache.ResourceEventHandlerFuncs{
+			AddFunc:    s.addEventHandler,
+			UpdateFunc: s.updateEventHandler,
+			DeleteFunc: s.deleteEventHandler,
+		},
+	)
+}
+
+func (s *sharedInformerHook) CreateWatch(name string, bufferSize int) <-chan watch.Event {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	ch := make(chan watch.Event, bufferSize)
+	s.objectNamesToWatch[name] = ch
+	return ch
+}
+
+func (s *sharedInformerHook) GetWatch(name string) <-chan watch.Event {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	return s.objectNamesToWatch[name]
+}
+
+func (s *sharedInformerHook) DeleteWatch(name string) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	delete(s.objectNamesToWatch, name)
+}
+
+func (s *sharedInformerHook) addEventHandler(obj interface{}) {
+	s.eventHandler(obj, watch.Added)
+}
+
+func (s *sharedInformerHook) updateEventHandler(oldObj, newObj interface{}) {
+	s.eventHandler(newObj, watch.Modified)
+}
+
+func (s *sharedInformerHook) deleteEventHandler(obj interface{}) {
+	s.eventHandler(obj, watch.Deleted)
+}
+
+func (s *sharedInformerHook) eventHandler(obj interface{}, eventType watch.EventType) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	metaObj, ok := obj.(metav1.Object)
+	if !ok {
+		return
+	}
+	runtimeObj, ok := obj.(runtime.Object)
+	if !ok {
+		return
+	}
+
+	if ch, ok := s.objectNamesToWatch[metaObj.GetName()]; ok {
+		ch <- watch.Event{
```

Comment:
I was unsure how to handle this.
- If the channel is unbuffered, then this event handler blocks until the channel is read
- If I write in a goroutine, then we risk leaking goroutines that are blocked until the channel is read
- If any of these buffered channels fill, then this event handler blocks everyone until it's read from

Any suggestions are appreciated.